routemaster/state_machine/utils.py

```diff
@@ -0,0 +1,188 @@
+    try:
+        return app.config.state_machines[label.state_machine]
+    except KeyError as k:
+        raise UnknownStateMachine(label.state_machine)
+
+
+def start_state_machine(app: App, label: LabelRef, conn) -> None:
+    """Create the first history entry for a label in a state machine."""
+    state_machine = get_state_machine(app, label)
+    conn.execute(history.insert().values(
+        label_name=label.name,
+        label_state_machine=label.state_machine,
+        old_state=None,
+        new_state=state_machine.states[0].name,
+    ))
+
+
+def choose_next_state(
+    state_machine: StateMachine,
+    current_state: State,
+    context: Context,
+) -> State:
+    """Assuming a transition is valid, choose a next state."""
+    next_state_name = current_state.next_states.next_state_for_label(context)
+    return state_machine.get_state(next_state_name)
+
+
+def get_label_metadata(
+    label: LabelRef,
+    state_machine: StateMachine,
+    conn,
+) -> Tuple[Dict[str, Any], bool]:
+    """Get the metadata and whether the label has been deleted."""
+    return conn.execute(
+        select([labels.c.metadata, labels.c.deleted]).where(and_(
+            labels.c.name == label.name,
+            labels.c.state_machine == state_machine.name,
+        )),
+    ).fetchone()
+
+
+def get_current_state(
+    label: LabelRef,
+    state_machine: StateMachine,
+    conn,
+) -> State:
+    """Get the current state of a label, based on its last history entry."""
+    history_entry = conn.execute(
+        select([history]).where(and_(
+            history.c.label_name == label.name,
+            history.c.label_state_machine == state_machine.name,
+        )).order_by(
+            history.c.created.desc(),
+        ).limit(1)
+    ).fetchone()
+
+    if history_entry is None:
+        raise UnknownLabel(label)
+
+    return state_machine.get_state(history_entry.new_state)
+
+
+def needs_gate_evaluation_for_metadata_change(
+    state_machine: StateMachine,
+    label: LabelRef,
+    update: Metadata,
+    conn,
+) -> bool:
+    """
+    Given a change to the metadata, should the gate evaluation be triggered.
+    """
+
+    current_state = get_current_state(label, state_machine, conn)
+
+    if not isinstance(current_state, Gate):
+        # Label is not a gate state so there's no trigger to resolve.
+        return False
+
+    if any(
+        trigger.should_trigger_for_update(update)
+        for trigger in current_state.metadata_triggers
+    ):
+        return True
+
+    return False
+
+
+def lock_label(label: LabelRef, conn):
+    """Lock a label in the current transaction."""
+    row = conn.execute(
+        labels.select().where(
+            and_(
+                labels.c.name == label.name,
+                labels.c.state_machine == label.state_machine,
+            ),
+        ).with_for_update()
+    ).fetchone()
+
+    if row is None:
+        raise UnknownLabel(label)
+
+    return row
+
+
+def labels_to_retry_for_action(
```

Comment:
This util looks complex enough to me to warrant directly unit testing. I think that might be partly that I'm not confident in reading the sqlalchemy bits, though I'd hope adding the tests would be easy.

Note that I think we'll want a general version of this which is `labels_in_state(state_machine, state_name, conn)` for use with time-triggered gates at some point, though that can come later.

Reply:
Yep that sounds like a good evolution of this. Do you think we could do the tests of that after we've done the test refactor for creating users in states? I think it would be much easier and nicer.

Reply:
Hrm, good point. Can we add a TODO or something here so we don't forget to come back and add those tests?

Reply:
I will add to the Asana list.